### PR TITLE
Improve shell scripts

### PR DESCRIPTION
**Pull request by redianthus**
Hi,

I fixed some stuff spotted using: `find -O3 . -type f -name '*.sh' -exec shellcheck -x  {} \;`.

Some warnings are remaining:

```sh
In ./.travis-ci.sh line 59:
      -with-instrumented-runtime $CONFIG_ARG
                                 ^-- SC2086: Double quote to prevent globbing and word splitting.


In ./.travis-ci.sh line 63:
      -with-instrumented-runtime $CONFIG_ARG \
                                 ^-- SC2086: Double quote to prevent globbing and word splitting.


In ./appveyor_build.sh line 68:
    echo 'eval $($APPVEYOR_BUILD_FOLDER/tools/msvs-promote-path)' >> ~/.bash_profile
         ^-- SC2016: Expressions don't expand in single quotes, use double quotes for that.
```

I believe the first and the second are false positives, as I guess that `$CONFIG_ARG` contains many arguments. We could instead use an array which is probably the best solution, or juste leave it like this.

For the third, it's definitively a false positive.

I can also add a comment like: `# shellcheck disable=SCxxxx` before each of them, to specifically disable these checks on these lines. Which would ease the process of checking again for some issues later.

Cheers,

### CODE REVIEW

File: .travis-ci.sh
Comment (dra27) [inline .travis-ci.sh]: This change is unnecessary - the entire script is running via `bash -e` and this directory will exist anyway - no one is going to delete the entire testsuite and that alteration go otherwise unnoticed!

File: .travis-ci.sh
Comment (dra27) [inline .travis-ci.sh]: I suppose, although `$(pwd)` in a Travis setup is highly unlikely to start containing spaces suddenly.

File: appveyor_build.sh
Comment (dra27) [inline appveyor_build.sh]: I think this is technically better, yes.

File: appveyor_build.sh
Comment (dra27) [inline appveyor_build.sh]: What is the purpose in splitting the `export` and assignment?

File: appveyor_build.sh
Comment (dra27) [inline appveyor_build.sh]: This script is run with `bash -e`

File: driver/ocamlcomp.sh.in
Comment (dra27) [inline driver/ocamlcomp.sh.in]: Unless the script explicitly uses bash, `$(` is relevantly less portable than back-ticks (some Solaris users, csh users, etc.)

**Review by dra27 — COMMENTED** [review]
I'm afraid I'm really not convinced of the value of this. In particular, changing back-ticks to `$(` has portability concerns (even if they're diminishing) and many of these double-quotings are around variables which can simply *never* contain spaces. AppVeyor, in particular, already tortures both scripts and build system with directories with spaces.

File: appveyor_build.sh
Comment (redianthus) [reply appveyor_build.sh]: See [SC2155](https://github.com/koalaman/shellcheck/wiki/SC2155).

File: driver/ocamlcomp.sh.in
Comment (redianthus) [reply driver/ocamlcomp.sh.in]: I think it's the opposite, see [SC2006](https://github.com/koalaman/shellcheck/wiki/SC2006) and the related resources section.

File: .travis-ci.sh
Comment (redianthus) [reply .travis-ci.sh]: In `.travis.yml` it's better to use:

```yml
script: ./.travis.sh
```

And so, to add `set -ex` in the `.sh` script. Will update this.

File: appveyor_build.sh
Comment (redianthus) [reply appveyor_build.sh]: Should be moved to this script then. As for the travis one. Updated.

File: .travis-ci.sh
Comment (gasche) [inline .travis-ci.sh]: Is this an intended change or was it done for testing purpose?

**Review by gasche — COMMENTED** [review]
For a minor countertone on it, I generally like the idea of using due diligence on tooling-assisted quality analysis (QA) for the codebase, including its shell scripts. (I think having repeatable automated QA tests has some value, which might compensate the relative lack of interest in some of the changes.) On the other hand, I'm not the one who is able to contribute meaningful review work to these change proposals, so I think the opinion of @dra27 matters more.

File: .travis-ci.sh
Comment (redianthus) [reply .travis-ci.sh]: It's intended. It was originally in the `.travis.yml` (see the diff of this file) file, but it's better to have it here.

File: driver/ocamlcomp.sh.in
Comment (pmetzger) [reply driver/ocamlcomp.sh.in]: $() is fully portable to all POSIX shells. No OS has deployed a shell that does not understand it in over 20 years.

File: .travis-ci.sh
Comment (dra27) [reply .travis-ci.sh]: This requires the scripts to be invoked using the execute bit, which is intentionally not the case at the moment.

File: appveyor_build.sh
Comment (dra27) [reply appveyor_build.sh]: I sort of end up repeating the question, I'm afraid!

File: .travis-ci.sh
Comment (dra27) [reply .travis-ci.sh]: Oh no, having reviewed the changes again, I see you only did this for the Travis script, not AppVeyor.

File: driver/ocamlcomp.sh.in
Comment (dra27) [reply driver/ocamlcomp.sh.in]: @zapashcanon - in the contexts in which backticks are being used, they work. Changing it brings none of the benefits (because we don't need to nest them and we don't hit the escaping problems, and so forth) but it does break the scripts for *some* users.

@pmetzger - that's simply not true. Solaris 10's `sh` does not understand this syntax, and it's a tier 2 platform.

Comment (pmetzger) [reply driver/ocamlcomp.sh.in]: Solaris 10's sh isn't POSIX conformant?

Comment (dra27) [reply driver/ocamlcomp.sh.in]: I don't have access to Solaris 10, but according to [this](https://unix.stackexchange.com/a/218073) and other tickets say this is so.

Note that /bin/sh can be non-POSIX but the system still be POSIX-compliant.

Comment (pmetzger) [reply driver/ocamlcomp.sh.in]: I'd forgotten about the entire Solaris mess. It was the only real exception to the rule.

Comment (dra27) [reply driver/ocamlcomp.sh.in]: Sure - it may be the only real exception, but it is *an* exception and that for me is enough not to change it, given that as far as I'm aware, there is no need to beyond linting. (I do add that I of course use `$()` in anything new and only use backticks in error, or interactively as it's one character fewer)

Comment (pmetzger) [reply driver/ocamlcomp.sh.in]: @dra27 I don't disagree. If portability to Solaris 10 is needed, then this is needed.

**Comment by shindere** [comment]
Perry E. Metzger (2018/04/02 12:54 +0000):
> pmetzger commented on this pull request.
> 
> 
> 
> > @@ -15,6 +15,6 @@
>  #*                                                                        *
>  #**************************************************************************
>  
> -topdir=`dirname $0`
> +topdir=$(dirname $0)
> 
> @dra27 I don't disagree. If portability to Solaris 10 is needed, then
> this is needed.

It'd be nice to be able to configure the linter not to warn about these
constructs,that would make it possible to use the tool for QA as @gasche
suggested and perhaps catch other potential problems, now or later.

**Comment by dra27** [comment]
@gasche - sort of, yes, although the problem with shell scripting is that these are linting tools, so many of the tests are highly subjective. I personally find that quoting everywhere implies less thought by the author than proper consideration of whether quoting is *necessary*. For example, inspection (and even static analysis, if one assumes that `sed` really means `sed`) of `tools/make-version-header.sh` should show that `$major`, `$minor` and `$patchlevel` can never contain spaces and so never need to be quoted. `$suffix` on the other hand (the optional portion after `+` in `VERSION`) is an arbitrary string and, as a result, is correctly quoted everywhere.

If I may address these changes in four parts:

## Shell eval quoting (i.e. `$()` vs back-ticks)

I haven't got immediate access to Solaris to verify this, but if Solaris 9 and 10 aren't presently broken building OCaml then, from what I can find online, making this change to the eval quoting will break them and so I don't see the benefit in this change. Alternatively, the change should be accompanied with the POSIX-recommended `#!/usr/bin/env sh` alteration - but that's also a high-risk change.

## Other changes to build system scripts (i.e. non-Travis and non-AppVeyor-related)

There's the quoting of a `-I` in `driver/ocamlcomp.sh.in` and the quoting of the filename passed to `sed` in `tools/make-version-header.sh`. Those changes aren't relevant to the build system because of how those scripts are invoked, but I agree that they're sound changes.

## Travis scripts

For the CI scripts, which are running very regularly (i.e. which don't have a habit of rotting) and which are also highly specific), I'd operate a much lower threshold for QA - I care much more about how quickly they can be inspected when being altered, than whether they pass a linting tool.

I don't find these quoting changes necessary, but I can live with them. The only exceptions are a few cases of unnecessary additional quoting (e.g. `"$TRAVIS_MERGE_BASE..$TRAVIS_PR_HEAD"` rather than `"$TRAVIS_MERGE_BASE..$TRAVIS_PR_HEAD"`)

## AppVeyor scripts

Again, I find quoting variables which cannot possibly contains spaces (e.g. the flexdll installation loop) just adds unnecessary noise to already tedious scripts and there are various others which, as far as I'm aware, can similarly never matter (granted, `$PROGRAMFILES` could be `C:\Program  Files` - two spaces, so squashes by word splitting if not quoted - on a truly twisted system). The splitting of the export on the grounds of an unlikely failure of `cygpath` (which would be detected very early in the build process anyway) to me adds more brittleness, at least in terms of moving the line around, as it's crucial that the two lines stay together. But, meh, linting...

**Comment by xavierleroy** [comment]
It's fun to see so much debate around shell scripts!  My two cents: 
* Double-quoting $VARIABLE expansions is a safe default.  You never know whose maniac is going to put spaces into this variable, even if and especially if it is a file path.  (Remember this popular operating system that puts most of its files into a directory called `Program Files` ?).  Not quoting should be the exception (when the variable actually needs to be split in words) rather than the rule.
* Backticks are OK the way we use them, so let's not get into a war concerning `$(...)`.

File: appveyor_build.sh
Comment (stedolan) [reply appveyor_build.sh]: IIUC, `export PATH=$(...)` will continue if `cygpath` explodes, despite `set -e`, while separate assignment and export will quit.

Comment (dra27) [reply appveyor_build.sh]: Indeed, that's my understanding - but if cygpath explodes, there will be serious trouble shortly, including but probably not limited to the lack of flexlink in PATH. These failures will be obvious - accidentally moving the `PATH=` elsewhere and forgetting to move the `export` line with it on the other hand is a bit more subtle to figure out what you did, amidst long waits in the AppVeyor build queue!

**Comment by mshinwell** [comment]
To add another opinion on this: if there is a reliable tool available for checking shell scripts, which it seems there might be, then I think the benefits of using such a tool (and following its recommendations) far outweigh minor disadvantages such as not being able to build on Solaris 10 or earlier.  I'm surprised there is so much pushback on this.  Even CI scripts waste people's time when they go wrong, after all...

As for the backticks, as far as I know, it is the modern recommendation to use dollar-parentheses instead as they are overall more robust.  I don't think it's really a strong argument to claim correctness based on our current usage, since that might easily change, and as we all know it can be really hard to spot such changes in shell language.

**Comment by gasche** [comment]
Note: the discussion may suggest that using dollars would break the Solaris build, but I think this is not what @dra27 meant: we can probably ensure that the concerned script runs with `bash` rather than `sh` and we could have both the dollars and the sun.

File: appveyor_build.sh
Comment (stedolan) [reply appveyor_build.sh]: Hmm, thinking about this some more I don't think the `export` line is needed at all. `PATH` is always exported, so I think ordinary assignments will show up in the environment.

**Comment by redianthus** [comment]
I rebased against current trunk and I changed the shebang in the files using `$()` to bash. Is there anything else that I should change ?

> Not quoting should be the exception (when the variable actually needs to be split in words) rather than the rule.

Agree. And in case where the variable actually needs to be split, you can most of the time switch to an array.

File: tools/make-version-header.sh
Comment (dra27) [inline tools/make-version-header.sh]: Doesn't this break on at least one BSD? (@avsm; @hannesm?)

Comment (hannesm) [reply tools/make-version-header.sh]: yes, this is a bad idea. better use `#!/usr/bin/env bash` if you really need a `bash`, i.e. when a POSIX sh is not sufficient (AFAICT the shell scripts being modified by this PR are all POSIX sh compliant, there's no need to invoke `bash`).

Comment (pmetzger) [reply tools/make-version-header.sh]: Many BSDs do not install bash by default, period, so there's no bash to be found.

Maybe this has proven to be more trouble than it is worth...

**Comment by redianthus** [comment]
Everything should be fine now. I removed usage of `$()` in scripts using `#!/bin/sh`.

File: .travis-ci.sh
Comment (dra27) [inline .travis-ci.sh]: The middle double-quotes are unnecessary - could this be changed to `"$TRAVIS_MERGE_BASE..$TRAVIS_PR_HEAD"`, please

File: .travis-ci.sh
Comment (dra27) [inline .travis-ci.sh]: Same as a previous - `"$TRAVIS_MERGE_BASE..$TRAVIS_PR_HEAD"`

File: appveyor_build.sh
Comment (dra27) [inline appveyor_build.sh]: @stedolan is correct, the `export` is in fact completely unnecessary (see [Posix](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_05_03)).

File: .travis-ci.sh
Comment (damiendoligez) [inline .travis-ci.sh]: Please pick one style and stick to it. My preference goes to the style used in line 145 of this file:
`"$TRAVIS_MERGE_BASE..$TRAVIS_PR_HEAD"`

File: .travis-ci.sh
Comment (damiendoligez) [inline .travis-ci.sh]: Same remark as above.

**Review by dra27 — CHANGES_REQUESTED** [review]
Thanks for the changes - just a couple of nits remain.

File: appveyor_build.sh
Comment (damiendoligez) [reply appveyor_build.sh]: Did the lint tool report a warning on this line? If not, I'm definitely siding with @dra27 here.

@stedolan you're right, unless somebody is foolish enough to do `export -n PATH`...

**Comment by dra27** [comment]
Regarding the `$CONFIG_ARG` false positives. Given that this is only used to pass the `-flambda` flag, it could be changed in `.travis.yml` to be `MIDDLEEND=-flambda` and then used quoted in `.travis-ci.sh`

I struggle to see why warning `SC2016` would ever be useful, given that the whole point of using single quotes is not to expand things!

Would you be up for adding the `shellcheck` as one of the Travis checks?

**Review by damiendoligez — CHANGES_REQUESTED** [review]
Looks good in general, and I agree that quoting should be the default.

The `export PATH=` business should be sorted out before we merge this.

File: .travis-ci.sh
Comment (redianthus) [reply .travis-ci.sh]: Done.

File: .travis-ci.sh
Comment (redianthus) [reply .travis-ci.sh]: Done.

File: appveyor_build.sh
Comment (redianthus) [reply appveyor_build.sh]: I'm OK with removing it. Can you point to the relevant part of your link (I try to find it but I failed as there's a lot of thing in it...) ? Just out of curiosity.

**Comment by redianthus** [comment]
> Regarding the $CONFIG_ARG false positives. Given that this is only used to pass the -flambda flag, it could be changed in .travis.yml to be MIDDLEEND=-flambda and then used quoted in .travis-ci.sh

The correct way would be to use an array I guess.

> I struggle to see why warning SC2016 would ever be useful, given that the whole point of using single quotes is not to expand things!

It's actually only because that's a mistake people often do, see the [wiki](https://github.com/koalaman/shellcheck/wiki/SC2016). We can just disable it.

> Would you be up for adding the shellcheck as one of the Travis checks?

Yes sure. Should I do it in a separate PR ?

File: .travis-ci.sh
Comment (redianthus) [reply .travis-ci.sh]: Done.

File: .travis-ci.sh
Comment (redianthus) [reply .travis-ci.sh]: Done.

**Review by damiendoligez — APPROVED** [review]

File: appveyor_build.sh
Comment (dra27) [reply appveyor_build.sh]: It's the sentence "If a variable is initialized from the environment, it shall be marked for export immediately" at http://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_05_03

You can see this if you have two scripts `foo` and `bar`:

`foo`:
```sh
#!/bin/sh
FOO='Updated FOO'
./bar
```
`bar`:
```sh
#!/bin/sh
echo "FOO=$FOO"
```
<pre><font color="#8AE234"><b>dra@cosmic</b></font>:<font color="#729FCF"><b>~</b></font>$ ./foo
FOO=
<font color="#8AE234"><b>dra@cosmic</b></font>:<font color="#729FCF"><b>~</b></font>$ FOO=foo ./foo
FOO=Updated FOO
</pre>

**Review by dra27 — APPROVED** [review]

**Comment by dra27** [comment]
Apologies for more than a year's silence on this PR. I have rebased your commit, taking into account a few new quotings in the spirit of what you did. I've added a commit to the bash scripts to standardise on `[[..]]` instead of `test` or `if [..]` which suppresses quite a few shellcheck warnings on double-quoting in a way which is IMO neater. Ignoring `build-aux` (which is entirely third party code, so we have no interest in changing it) and, for now, `testsuite`, the following warnings remain:

<pre><b>In ./tools/ci/travis/travis-ci.sh line 99:</b>
    ./configure $configure_flags
<font color="#4E9A06">                ^-- SC2086: Double quote to prevent globbing and word splitting.</font>

<b>In ./tools/ci/travis/travis-ci.sh line 104:</b>
      $configure_flags
<font color="#4E9A06">      ^-- SC2086: Double quote to prevent globbing and word splitting.</font></pre>

These two warnings are incorrect - a contributor with more time may like to convert it to use bash arrays instead - I don't think it should be hard-coded to accept only one parameter (i.e. just quote it) because that's likely to be very annoying when someone in future wishes to put more than one parameter in.

<pre><b>In ./tools/ci/travis/travis-ci.sh line 158:</b>
  elif [[ -n $(curl &quot;$API_URL&quot; | grep &apos;no-change-entry-needed&apos;) ]]
<font color="#4E9A06">          ^-- SC2143: Use grep -q instead of comparing output with [ -n .. ].</font></pre>

Pfff! Sure, filed under "more inclined than me"!

<pre><b>In ./tools/ci/appveyor/appveyor_build.sh line 87:</b>
      echo &apos;eval $($APPVEYOR_BUILD_FOLDER/tools/msvs-promote-path)&apos; \
<font color="#4E9A06">           ^-- SC2016: Expressions don&apos;t expand in single quotes, use double quotes for that.</font></pre>

This warning is incorrect and, as @mshinwell notes, I think would be better suppressed completely if we were to use this regularly.

<pre><b>In ./tools/ci/appveyor/appveyor_build.sh line 120:</b>
      cd ..
<font color="#4E9A06">      ^-- SC2103: Use a ( subshell ) to avoid having to cd back.</font></pre>

Pfff! I think using a subshell to achieve this is poor coding, to be honest. I guess we could use `pushd` and `popd`, but I find having to write `pushd .. &>/dev/null` tedious.

<pre><b>In ./manual/tools/fix_index.sh line 43:</b>
    -e &apos;s/verb`(&quot;|hyperindexformat{\\&quot;}/verb`(&quot;|&quot;|)`|hyperpage/&apos; \
<font color="#4E9A06">       ^-- SC2016: Expressions don&apos;t expand in single quotes, use double quotes for that.</font>

<b>In ./manual/tools/fix_index.sh line 44:</b>
    -e &apos;s/verb`(&quot;|hyperindexformat{\\&gt;)`}/verb`(&quot;|&gt;)`|hyperpage/&apos;
<font color="#4E9A06">       ^-- SC2016: Expressions don&apos;t expand in single quotes, use double quotes for that.</font></pre>

As with the previous instance of this warning, this warning again demonstrates its ill-conceivedness! Perhaps `shellcheck` should have an `--i-am-not-completely-green` parameter :wink:

<pre><b>In ./runtime/gen_primitives.sh line 28:</b>
  sed -n -e &apos;s/^CAMLprim_int64_[0-9](\([a-z0-9_][a-z0-9_]*\)).*/caml_int64_\1\
<font color="#4E9A06">                                                                             ^-- SC1004: This backslash+linefeed is literal. Break outside single quotes if you just want to break the line.</font></pre>

This is incorrect - it's actually trying to pass backslash and a linefeed literal to `sed`. Given that this is one of those "this really must work on any `sh` you can conceive" scripts, I'd live with this one and add the appropriate magic comments to suppress that warning if `shellcheck` becomes part of CI.

I'd be happy to see `shellcheck` used in CI (and also `checkbashisms`, which could be hooked into the check-typo logic on Travis), but whoever contributes that will need to address these warnings, preferably as I note here, and also bring the testsuite scripts into compliance.

**Comment by dra27** [comment]
Obviously, the CI scripts are somewhat self-checking. The changes in the other scripts look safe to me - @damiendoligez, would you like to cast an eye over them too and merge if you're happy?

**Comment by redianthus** [comment]
Thanks. I forgot about this PR...

> I guess we could use pushd and popd

Please don't. It's not POSIX. And IIRC, shellcheck has some kind of warning for them for some other reason.

About adding shellcheck to the CI, maybe I'll do it soon, if this first PR gets merged, and fix the double-quote/array thing and others you mentionned.

**Comment by dra27** [comment]
The fact `pushd` and `popd` aren't POSIX isn't relevant - it's in a bash script (I did initially try it, and `shellcheck` passes it) however, I just think it's less clear than what was there before in that specific place!

**Comment by damiendoligez** [comment]
@dra27 Thanks for picking up the dropped ball.

**Pull request merged**
